File: android/java/app/src/main/java/com/daily/partyline/Participant.java
package com.daily.partyline;

public class Participant {

    public static String myId;
    public static final String SPEAKER_TAG = "SPK";
    public static final String RAISE_HAND_TAG = "✋ ";
    public static final String LISTENER_TAG = "LST";
    public static final String MODERATOR_TAG = "MOD";
    private String mUserName;
    private String mId;
    private Boolean mIsModerator;
    private Boolean mIsSpeaker;
    private Boolean mIsActiveSpeaker;
    private Boolean mIsHandRaised;
    private Boolean mIsMuted;

    public Participant(String userName, String Id, Boolean isModerator) {
        mUserName = userName.replace("_", "");
        mId = Id;
        mIsMuted = true;
        mIsModerator = isModerator;
        mIsSpeaker = isModerator || userName.contains(SPEAKER_TAG);
        mIsActiveSpeaker = false;
        mIsHandRaised = userName.contains(RAISE_HAND_TAG);
    }

    public void update(String userName, Boolean isModerator) {
        mIsModerator = isModerator;
        setUserName(userName);
    }

    public String getUserName() {
        return mUserName;
    }

    public String getId() {
        return mId;
    }

    public Boolean getIsModerator() {
        return mIsModerator;
    }

    public Boolean getIsSpeaker() {
        return mIsSpeaker;
    }

    public void setUserName(String userName) {
        mUserName = userName;//.replace("_", "");
        mIsSpeaker = mIsModerator || userName.contains(SPEAKER_TAG);
        mIsHandRaised = userName.contains(RAISE_HAND_TAG);
    }

    public void setId(String mId) {
        mId = mId;
    }

    public void setIsModerator(Boolean isModerator) {
        mIsModerator = isModerator;
    }

    public void setIsSpeaker(Boolean isSpeaker) {
        setUserName(isSpeaker ? mUserName.concat(SPEAKER_TAG) : mUserName.replace(SPEAKER_TAG, ""));
    }

    public Boolean getIsHandRaised() {
        return mIsHandRaised;
    }

    public void setIsHandRaised(Boolean raiseHand) {
        setUserName(raiseHand ? RAISE_HAND_TAG.concat(mUserName) : mUserName.replace(RAISE_HAND_TAG, ""));
        mIsHandRaised = raiseHand;
    }

    public Boolean getIsActiveSpeaker() {
        return mIsActiveSpeaker;
    }

    public void setIsActiveSpeaker(Boolean isActiveSpeaker) {
        mIsActiveSpeaker = isActiveSpeaker;
    }

    public Boolean getIsMuted() {
        return mIsMuted;
    }

    public void setIsMuted(Boolean isMuted) {
        mIsMuted = isMuted;
    }

    public void cleanUsername() {
        mUserName = mUserName.replace("_", "");
        mUserName = mUserName.replace(Participant.MODERATOR_TAG, "");
        mUserName = mUserName.replace(Participant.LISTENER_TAG, "");
        mUserName = mUserName.replace(Participant.SPEAKER_TAG, "");
    }
}

File: android/kotlin/app/src/main/java/com/daily/partyline/Participant.kt
package com.daily.partyline

class Participant(userName: String, Id: String, isModerator: Boolean) {
    private var userName: String
    private val id: String
    private var isModerator: Boolean
    private var isSpeaker: Boolean
    private var isActiveSpeaker: Boolean
    private var isHandRaised: Boolean
    private var isMuted: Boolean
    fun update(userName: String, isModerator: Boolean) {
        this.isModerator = isModerator
        setUserName(userName)
    }

    fun getUserName(): String? {
        return userName
    }

    fun getId(): String? {
        return id
    }

    fun getIsModerator(): Boolean? {
        return isModerator
    }

    fun getIsSpeaker(): Boolean? {
        return isSpeaker
    }

    fun setUserName(userName: String) {
        this.userName = userName//.replace("_", "")
        isSpeaker = isModerator == true || userName.contains(SPEAKER_TAG.toString())
        isHandRaised = userName.contains(RAISE_HAND_TAG.toString())
    }

    fun setId(mId: String?) {
        var mId = mId
        mId = mId
    }

    fun setIsModerator(isModerator: Boolean) {
        this.isModerator = isModerator
    }

    fun setIsSpeaker(isSpeaker: Boolean) {
        setUserName(if (isSpeaker) userName + SPEAKER_TAG else userName.replace(SPEAKER_TAG.toString(), ""))
    }

    fun getIsHandRaised(): Boolean? {
        return isHandRaised
    }

    fun setIsHandRaised(raiseHand: Boolean) {
        setUserName(if (raiseHand) RAISE_HAND_TAG + userName else userName.replace(RAISE_HAND_TAG.toString(), ""))
        isHandRaised = raiseHand
    }

    fun getIsActiveSpeaker(): Boolean? {
        return isActiveSpeaker
    }

    fun setIsActiveSpeaker(isActiveSpeaker: Boolean) {
        this.isActiveSpeaker = isActiveSpeaker
    }

    fun getIsMuted(): Boolean? {
        return isMuted
    }

    fun setIsMuted(isMuted: Boolean) {
        this.isMuted = isMuted
    }

    fun cleanUsername() {
        userName = userName?.replace("_", "")
        userName = userName?.replace(MODERATOR_TAG.toString(), "")
        userName = userName?.replace(LISTENER_TAG.toString(), "")
        userName = userName?.replace(SPEAKER_TAG.toString(), "")
    }

    companion object {
        var myId: String? = null
        val SPEAKER_TAG: String = "SPK"
        val RAISE_HAND_TAG: String = "✋ "
        val LISTENER_TAG: String = "LST"
        val MODERATOR_TAG: String = "MOD"
    }

    init {
        this.userName = userName.replace("_", "")
        id = Id
        isMuted = true
        this.isModerator = isModerator
        isSpeaker = isModerator || userName.contains(other = SPEAKER_TAG.toString())
        isActiveSpeaker = false
        isHandRaised = userName.contains(other = RAISE_HAND_TAG.toString())
    }
}